### ISSUE TITLE
find a better approach to simplify polygons

### ISSUE THREAD

**Issue by ltclm** [geoadmin/vector-forge#38]
As we recently experienced the current approach to simplify polygon geometries with common boundaries is not ideal.
The following approach could be worth a try [1].
After the integration of the data we have to build a topology with postgis:
```sql
SELECT CreateTopology('topology_karto_dkm10', 21781, 0.5);
SELECT AddTopoGeometryColumn('topology_karto_dkm10', 'karto', 'dkm10_bodenbedeckung', 'the_geom_topo', 'MULTIPOLYGON');
UPDATE karto.dkm10_bodenbedeckung SET the_geom_topo = toTopoGeom(the_geom, 'topology_karto_dkm10', 1 , 0.5);
```

The new topoGeometry can be simplified on the fly with the desired tolerance:
```sql
SELECT objectid, st_simplify(the_geom_topo,5)  as the_geom
  FROM karto.dkm10_bodenbedeckung WHERE the_geom && st_makeEnvelope(600000,200000,630000,230000,21781);
``` 

Original Geometry:
![image](https://cloud.githubusercontent.com/assets/5286659/15213389/29884300-1847-11e6-820f-08f51086bc9e.png)

st_simplify with original geometry:
![image](https://cloud.githubusercontent.com/assets/5286659/15213408/4685ae3e-1847-11e6-9f46-3f1a8b72b9a5.png)

st_simplify with topoGeometry
![image](https://cloud.githubusercontent.com/assets/5286659/15213421/5b786354-1847-11e6-8d70-12e4c8a9b343.png)

[1] https://strk.kbt.io/blog/2012/04/13/simplifying-a-map-layer-using-postgis-topology/

**Comment by gjn** [comment]
Awsome!

**Comment by loicgasser** [comment]
:rocket:
Sure, have you prepared a DB with bodenbedeckung I can use?
I'll add on the fly simplification based on your query samples.

**Comment by ltclm** [comment]
i will update stopo_master tonight

**Comment by loicgasser** [comment]
FYI we have 1'000 times more vertices than in the above example. Might still take a while :)

**Comment by ltclm** [comment]
yes, i have written the following script to parallelize the update:
```bash
#!/bin/bash

for i in {0..7}; do
    psql -c "begin;SET work_mem='1GB';UPDATE karto.dkm10_bodenbedeckung SET the_geom_topo = toTopoGeom(the_geom, 'topology_karto_dkm10', 1 , 0.5) where the_geom_topo is null order by objectid limit 64296 offset ${i}*64296;commit;" -d stopo_master &> /tmp/update_${i}.log &
done

```

**Comment by loicgasser** [comment]
I red the article and I have a comment to make on `SimplifyEdgeGeom`

It says:

> The SimplifyEdgeGeom function wraps the ST_ChangeEdgeGeom call into a subtransaction and handles exceptions by reducing the simplification factor until it succeeds. The version I used reduces simplification factor in half at each failure, dropping down to zero around 1e-8. You can roll your own with other heuristics or generalize this one to take parameters about stepping and limits.

I think that instead of reducing the simplification factor, we should write empty geometries we can later filter.
Or at least be able to decide whether we want to simplify (reduce the number of points) and generalize (remove geometries that are not needed because they are too small to display) or only simplify.

What do you think?

**Comment by ltclm** [comment]
Hi Loic

The import is still ongoing, it will be finished this late afternoon. There is a btree index on the new column ``the_geom_topo``. Features which could not be added to the topology have a NULL value in this column, so you can filter them out. spatial intersections have to be done on the original geom column. 

I think we should do both, bbox/area based filter would make sense in my eyes. we have to create an index for that too.

**Comment by loicgasser** [comment]
Awesome ! I already filter by empty geometry collection. I ll adapt as needed. Could you commit the sql and make a PR ? I ll make a new run this week-end. Is it on stop_tileforge_master?

**Comment by ltclm** [comment]
Hi Loic

Building the topology took much longer than excpected...
The data is now ready. You will find a new attribute ``the_geom_topo`` in the table ``stopo_master.karto.dkm10_bodenbedeckung``.
Spatial filters should be applied on the original geometry column ``the_geom``.
I was able to create the topology for 501919 out of 514374 features (97%). The column the_geom_topo has NULL value for the failed features.
Topology creation failed for different reasons.
Here the top 10 failures
```bash
$ grep "ERROR:" *.log | awk -F"ERROR:" '{print $2}' | sort | uniq -c | sort -k1nr | head -n 10
  10329   lwpoly_from_lwlines: shell must be closed
     89   deadlock detected
     13   Spatial exception - geometry intersects edge 672236
     10   Spatial exception - geometry intersects edge 81628
      9   Spatial exception - geometry intersects edge 91768
      9   SQL/MM Spatial exception - curve not simple
      6   Spatial exception - geometry intersects edge 286350
      6   SQL/MM Spatial exception - geometry crosses a node
      5   Spatial exception - geometry intersects edge 113494
      5   Spatial exception - geometry intersects edge 129960
```

You will find the import logs on ``10.220.5.186:/tmp/topo_insert_<<objectid>>.log``

To visualize the topology enabled data you have to call the following sql:
``SELECT *,the_geom_topo::geometry from karto.dkm10_bodenbedeckung;``

The script i used to for the topology creation is here:
https://github.com/geoadmin/vector-forge/pull/39

**Comment by loicgasser** [comment]
Nice, we should have a first version very soon. Thanks a lot for that.

**Comment by loicgasser** [comment]
Another useful link http://blog.mathieu-leplatre.info/use-postgis-topologies-to-clean-up-road-networks.html
